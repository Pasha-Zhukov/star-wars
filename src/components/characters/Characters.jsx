import { useCallback, useRef, useState } from "react";
import { Circles } from "../circles/Circles";
import { GenderCharacter } from "../genderCharacter/GenderCharacter";
import { Loader } from "../loader/Loader";
import { ModalWindow } from "../modalWindow/ModalWindow";
import "./Characters.scss";

export const Characters = ({
  setPageNum,
  charactersData,
  pageNum,
  isLoading,
  setIsModalActive,
  isModalActive,
}) => {
  const [dataCharacter, setDataCharacter] = useState("");

  const observer = useRef();
  const lastBookElementRef = useCallback(
    (node) => {
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting && pageNum < 9) {
          setPageNum((prev) => prev + 1);
        }
      });
      if (node) observer.current.observe(node);
    },
    [charactersData]
  );

  const handleClick = (e) => {
    setDataCharacter(
      charactersData.find((element) =>
        element.name
          ? element.name === e.currentTarget.dataset.name
          : element.whrascwo === e.currentTarget.dataset.name
      )
    );
    setIsModalActive(!isModalActive);
  };
  return (
    <>
      <div className="container">
        {charactersData?.map((item, index) => {
          const isLastItem = index === charactersData.length - 1;
          return (
            <div
              onClick={handleClick}
              className={`card-character ${isModalActive} ? "blur" : ""`}
              ref={isLastItem ? lastBookElementRef : undefined}
              key={item.name || item.whrascwo}
              data-name={item.name || item.whrascwo}
            >
              <div className="nameCharacter">{item.name || item.whrascwo}</div>
              <Circles item={item} isModalActive={isModalActive} />
              <GenderCharacter item={item} />
            </div>
          );
        })}
        <ModalWindow
          isModalActive={isModalActive}
          dataCharacter={dataCharacter}
          setIsModalActive={setIsModalActive}
          setDataCharacter={setDataCharacter}
        />
      </div>
      <div>{isLoading && <Loader />}</div>
    </>
  );
};
