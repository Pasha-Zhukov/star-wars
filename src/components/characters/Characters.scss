.container {
  padding: 0 156px;
  display: flex;
  flex-wrap: wrap;
}

.card-character {
  cursor: pointer;
  min-height: 136px;
  width: 351px;
  background: #f0f0f0;
  border-radius: 8px;
  margin: 15px 25px 30px 25px;
  padding: 5px 5px 5px 26px;
  filter: drop-shadow(4px 4px 4px rgba(33, 33, 33, 0.1));
  .nameCharacter {
    font-family: "Karla";
    font-style: normal;
    font-weight: 700;
    font-size: 18px;
    line-height: 21px;
    text-shadow: 4px 4px 4px rgba(33, 33, 33, 0.1);
    text-align: left;
    margin-bottom: 10px;
  }
}
@media (max-width: 930px) {
  .container {
    justify-content: center;
    padding: 0 30px;
  }
}
