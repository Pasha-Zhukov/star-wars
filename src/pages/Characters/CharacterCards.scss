* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}
.language {
  margin-right: 156px;
  margin-top: 28px;
  font-family: "Source Sans Pro";
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  align-items: center;
  text-align: right;
  color: #4b4b4b;
}
.title {
  font-family: "Karla";
  font-style: normal;
  font-weight: 400;
  font-size: 35px;
  line-height: 41px;
  text-align: center;
  letter-spacing: 3px;
  color: #000000;
  text-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  margin-bottom: 35px;
}
.block-select {
  margin-left: 175px;
  font-family: "Source Sans Pro";
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 20px;
  display: flex;
  align-items: center;
  text-align: center;
  color: #212121;

  .title-select {
    margin-right: 18px;
  }
  .select {
    height: 20px;
    min-width: 100px;
    text-align: center;
    border: none;
    border-radius: 4px;
    outline: none;
    background: #f0f0f0;
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    -webkit-appearance: button;
    -moz-appearance: button;
  }
}

.language-Button {
  cursor: pointer;
  height: 75px;
  width: 75px;
  position: fixed;
  bottom: 30px;
  right: 100px;
}
@media (max-width: 930px) {
  .language-Button {
    cursor: pointer;
    height: 55px;
    width: 55px;
    position: fixed;
    bottom: 20px;
    right: 20px;
  }
  .block-select {
    margin-left: 55px;
  }
}
