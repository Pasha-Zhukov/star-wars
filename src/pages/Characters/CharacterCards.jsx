import { Characters } from "../../components/characters/Characters";
import { Header } from "../../components/header/Header";
import { useState } from "react";
import language from "../../images/language.png";
import { useEffect } from "react";
import "./CharacterCards.scss";

export function CharacterCards() {
  const [charactersData, setCharactersData] = useState([]);
  const [charactersDataWookiee, setCharactersDataWookiee] = useState([]);
  const [currentLanguage, setCurrentLanguage] = useState("EN");

  const dataMap = new Map();
  dataMap.set("EN", charactersData);
  dataMap.set("WO", charactersDataWookiee);

  const [pageNum, setPageNum] = useState(1);
  const [isLoading, setIsLoading] = useState(false);
  const [selectColor, setSelectColor] = useState("all");
  const [isModalActive, setIsModalActive] = useState(false);

  const colorSet = [...new Set(charactersData.map((i) => i.eye_color))];
  const colorSetWookiee = [
    ...new Set(charactersDataWookiee.map((i) => i.worowo_oaooanoorc)),
  ];

  const onChangeLanguage = () => {
    setCurrentLanguage((prev) => (prev === "EN" ? "WO" : "EN"));
    setSelectColor((prev) => (prev === "all" ? "Whoh" : "all"));
    setPageNum(1);
    setCharactersData([]);
    setCharactersDataWookiee([]);
  };

  async function fetchPeople() {
    setIsLoading(true);
    let res = await fetch(
      `https://swapi.dev/api/people/?fotmat=json&page=${pageNum}`
    );
    res
      .json()
      .then((data) => {
        setCharactersData((prev) => [...prev, ...data.results]);
      })
      .finally(() => setIsLoading(false));
  }

  async function fetchPeopleWookiee() {
    setIsLoading(true);
    const res = await fetch(
      `https://swapi.dev/api/people/?format=wookiee&page=${pageNum}`
    );
    res
      .text()
      .then((text) => {
        text = text.replace(/whhuanan/g, '"whhuanan"');
        return JSON.parse(text);
      })
      .then((data) => {
        setCharactersDataWookiee((prev) => [...prev, ...data.rcwochuanaoc]);
      })
      .finally(() => setIsLoading(false));
  }
  useEffect(() => {
    if (isLoading) return;
    if (currentLanguage === "EN") {
      fetchPeople();
    } else fetchPeopleWookiee();
  }, [pageNum, currentLanguage]);

  return (
    <>
      <Header />
      <p className="language">language: {currentLanguage}</p>
      <h1 className="title">
        {charactersData?.length || charactersDataWookiee?.length}
        {currentLanguage === "EN"
          ? " Peoples for you to choose your favorite"
          : " rcwochuanaoc"}
      </h1>
      <div className="block-select">
        <label className="title-select">
          {currentLanguage === "EN" ? "color eye" : "worowo oaooanoorc"}
        </label>
        {currentLanguage === "EN" ? (
          <select
            onChange={(e) => setSelectColor(e.target.value)}
            name="nameChracter"
            className="select"
          >
            <option value="all">All</option>
            {colorSet?.map((color) => (
              <option value={color} key={color}>
                {color}
              </option>
            ))}
          </select>
        ) : (
          <select
            onChange={(e) => setSelectColor(e.target.value)}
            name="nameChracter"
            className="select"
          >
            <option value="Whoh">Whoh</option>
            {colorSetWookiee?.map((color) => (
              <option value={color} key={color}>
                {color}
              </option>
            ))}
          </select>
        )}
      </div>
      <Characters
        charactersData={
          selectColor === "all" || selectColor === "Whoh"
            ? dataMap.get(currentLanguage)
            : currentLanguage === "EN"
            ? dataMap
                .get(currentLanguage)
                .filter((item) => item.eye_color === selectColor)
            : dataMap
                .get(currentLanguage)
                .filter((item) => item.worowo_oaooanoorc === selectColor)
        }
        setPageNum={setPageNum}
        pageNum={pageNum}
        isLoading={isLoading}
        isModalActive={isModalActive}
        setIsModalActive={setIsModalActive}
      />
      <img
        onClick={onChangeLanguage}
        className="language-Button"
        src={language}
        alt="language"
      />
    </>
  );
}
