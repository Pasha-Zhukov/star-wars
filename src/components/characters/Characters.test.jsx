import { Characters } from "./Characters";
import renderer from "react-test-renderer";

const charactersData = [
  {
    birth_year: "19BBY",
    eye_color: "blue",
    gender: "male",
    hair_color: "blond",
    height: "172",
    mass: "77",
    name: "Luke Skywalker",
    skin_color: "fair",
  },
  {
    birth_year: "33BBY",
    eye_color: "red",
    gender: "n/a",
    hair_color: "n/a",
    height: "96",
    mass: "32",
    name: "R2-D2",
    skin_color: "white, blue",
  },
];

const handleClick = (e) => {
  return charactersData.find((element) =>
    element.name ? element.name === e : element.whrascwo === e
  );
};
describe("Name only requests", () => {
  test("valid name Person", () => {
    expect(handleClick("Luke Skywalker")).toStrictEqual({
      birth_year: "19BBY",
      eye_color: "blue",
      gender: "male",
      hair_color: "blond",
      height: "172",
      mass: "77",
      name: "Luke Skywalker",
      skin_color: "fair",
    });
  });
});
describe("Characters", () => {
  test("Get tree", () => {
    const tree = renderer.create(<Characters />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
