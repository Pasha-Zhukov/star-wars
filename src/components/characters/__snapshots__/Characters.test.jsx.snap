// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Get tree 1`] = `
Array [
  <div
    className="container"
  >
    <div
      className=""
      onClick={[Function]}
    />
    <div
      className="popup  "
    >
      <img
        alt="closeIcon"
        className="closeIcon"
        onClick={[Function]}
        src="closeIcon.png"
      />
      <div
        className="iconBlock"
      >
        <div
          className="wrapperModalGender"
        >
          <p
            className="birth-yearModal"
          />
        </div>
         
      </div>
      <div
        className="infoBlock"
      >
        <p
          className="nameModal"
        />
        <div
          className="colorsModal"
        >
          <div>
            null undefined
          </div>
          <div>
            null undefined
          </div>
          <div>
            null undefined
          </div>
        </div>
        <div
          className="wrapperModal"
        />
      </div>
    </div>
    <div />
  </div>,
  <div />,
]
`;
